dpctl/dparray.py

```diff
@@ -0,0 +1,257 @@
+                buffer=buf,
+                offset=0,
+                strides=strides,
+                order=order,
+            )
+            if hasattr(new_obj, array_interface_property):
+                dprint("buffer None new_obj already has sycl_usm")
+            else:
+                dprint("buffer None new_obj will add sycl_usm")
+                setattr(new_obj, array_interface_property, {})
+            return new_obj
+        # zero copy if buffer is a usm backed array-like thing
+        elif hasattr(buffer, array_interface_property):
+            dprint("dparray::ndarray __new__ buffer", array_interface_property)
+            # also check for array interface
+            new_obj = np.ndarray.__new__(
+                subtype,
+                shape,
+                dtype=dtype,
+                buffer=buffer,
+                offset=offset,
+                strides=strides,
+                order=order,
+            )
+            if hasattr(new_obj, array_interface_property):
+                dprint("buffer None new_obj already has sycl_usm")
+            else:
+                dprint("buffer None new_obj will add sycl_usm")
+                setattr(new_obj, array_interface_property, {})
+            return new_obj
+        else:
+            dprint("dparray::ndarray __new__ buffer not None and not sycl_usm")
+            nelems = np.prod(shape)
+            # must copy
+            ar = np.ndarray(
+                shape,
+                dtype=dtype,
+                buffer=buffer,
+                offset=offset,
+                strides=strides,
+                order=order,
+            )
+            nbytes = int(ar.nbytes)
+            buf = MemoryUSMShared(nbytes)
+            new_obj = np.ndarray.__new__(
+                subtype,
+                shape,
+                dtype=dtype,
+                buffer=buf,
+                offset=0,
+                strides=strides,
+                order=order,
+            )
+            np.copyto(new_obj, ar, casting="no")
+            if hasattr(new_obj, array_interface_property):
+                dprint("buffer None new_obj already has sycl_usm")
+            else:
+                dprint("buffer None new_obj will add sycl_usm")
+                setattr(new_obj, array_interface_property, {})
+            return new_obj
+
+    def __array_finalize__(self, obj):
+        dprint("__array_finalize__:", obj, hex(id(obj)), type(obj))
+        # When called from the explicit constructor, obj is None
+        if obj is None:
+            return
+        # When called in new-from-template, `obj` is another instance of our own
+        # subclass, that we might use to update the new `self` instance.
+        # However, when called from view casting, `obj` can be an instance of any
+        # subclass of ndarray, including our own.
+        if hasattr(obj, array_interface_property):
+            return
+        if isinstance(obj, numba.core.runtime._nrt_python._MemInfo):
```

Comment:
To wean `dpctl` off dependency on `numba` we need to rely on `obj` to provide `__sycl_usm_array_interface__` and make this the only way we detect that `obj` has the underlying USM memory.

We will also need to validate the content of the `__sycl_usm_array_interface__` making sure that the USM pointer is meaningful in the context that we work with.

Doing this will require changes in Numba-DPPY, and possibly Numba itself, which are queued up.

Reply:
The check on lines 120-121 should be `hasattr(obj, array_interface_property)` where `array_interface_property` is `__sycl_usm_array_interface__` and should supersede all the subsequent checks, but keep code on lines 142-148.

Reply:
Fixed on the dpctl side.